### PR TITLE
[SPARK-20164][SQL] AnalysisException not tolerant of null query plan.

### PR DESCRIPTION
**Pull request by kunalkhamar**
## What changes were proposed in this pull request?

The query plan in an `AnalysisException` may be `null` when an `AnalysisException` object is serialized and then deserialized, since `plan` is marked `@transient`. Or when someone throws an `AnalysisException` with a null query plan (which should not happen).
`def getMessage` is not tolerant of this and throws a `NullPointerException`, leading to loss of information about the original exception.
The fix is to add a `null` check in `getMessage`.

## How was this patch tested?

- Unit test

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala [hunk 1]
```diff
@@ -43,7 +43,7 @@ class AnalysisException protected[sql] (
   }
 
   override def getMessage: String = {
-    val planAnnotation = plan.map(p => s";\n$p").getOrElse("")
+    val planAnnotation = if (plan == null) "" else plan.map(p => s";\n$p").getOrElse("")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala]: Nit: 
```Scala
val planAnnotation = Option(plan).map(p => s";\n$p").getOrElse("")
```

**Comment by gatorsmile** [comment]
We should pass `None` for the parm `plan`, instead of `null`. We should fix the place we throw the `AnalysisException `.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala (hunk 1 above)
Comment (kunalkhamar) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala]: Changed.

**Comment by kunalkhamar** [comment]
@gatorsmile Looks like its happening because `plan` is marked `transient`, when AnalysisException is serialized and then deserialized, `plan` may turn up null.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala [hunk 2]
```diff
@@ -43,7 +43,7 @@ class AnalysisException protected[sql] (
   }
 
   override def getMessage: String = {
-    val planAnnotation = plan.map(p => s";\n$p").getOrElse("")
+    val planAnnotation = Option(plan).map(p => s";\n$p").getOrElse("")
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala]: Sorry, my fault. We need another `flatten`.
```Scala
    val planAnnotation = Option(plan).flatten.map(p => s";\n$p").getOrElse("")
```

Comment (kunalkhamar) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/AnalysisException.scala]: Right, updated.

**Comment by gatorsmile** [comment]
LGTM

It sounds like the support of the default value in Scala is weird. I did some searches. Scala omits the default value. Another alternative is to define auxiliary constructor function to explicitly set the default value. 

cc @rxin @hvanhovell @cloud-fan

**Comment by viirya** [comment]
LGTM

A variable marked with `@transient` will be restored to the default value of the type, not the default value of the parameter.

Quoted from Chapter 25 of Programming in Scala, First Edition: http://www.artima.com/pins1ed/annotations.html

> Finally, Scala provides a @transient annotation for fields that should not be serialized at all. If you mark a field as @transient, then the framework should not save the field even when the surrounding object is serialized. When the object is loaded, the field will be restored to the default value for the type of the field annotated as @transient.

**Comment by gatorsmile** [comment]
@kunalkhamar Could you summarize the issue and post it in https://github.com/databricks/scala-style-guide? Thanks!

**Comment by gatorsmile** [comment]
Thanks! Merging to master and 2.1.

**Comment by kunalkhamar** [comment]
@gatorsmile Verified the behaviour using this, it makes `plan` null upon deserialization.
```
import java.io._
import org.apache.spark.sql.AnalysisException

lazy val exception = new AnalysisException("", None, None, plan = None)
// Serialize exception
lazy val bo = new ByteArrayOutputStream()
lazy val o = new ObjectOutputStream(bo)
o.writeObject(exception)
lazy val bytes = bo.toByteArray

// Deserialize ex
lazy val bi = new ByteArrayInputStream(bytes)
lazy val i = new ObjectInputStream(bi)
lazy val deserialized = i.readObject.asInstanceOf[AnalysisException]

println(deserialized.plan)
```

Not sure what to add and where in the scala-style-guide?
